### PR TITLE
Support for GCS proxies everywhere in the GCS API

### LINKED ISSUE #91952: HTTP proxy settings are not consistently configured in the GCS client
### Elasticsearch Version

8.x (any version where we have added support for proxies with a GCS repo)

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

N/A

### Problem Description

When configuring a GCS repository with HTTP proxy settings ([doc](https://www.elastic.co/guide/en/elasticsearch/reference/8.0/repository-gcs.html#repository-gcs-client)), we do not pass along the proxy settings to other `HttpTransportOptions` instances that we create. This can lead to timeout errors when the GCS Client tries to get a token, with the following error message:
```
Error getting access token for service account: Connect timed out, iss: ...
```

The workaround is to set the HTTP proxy settings on the JVM level rather than when defining the GCS repository, using `java.net.useSystemProxies` or `http.proxyHost`/`http.proxyPort`.

### Steps to Reproduce

[Setting up a GCS repository](https://www.elastic.co/guide/en/elasticsearch/reference/8.0/repository-gcs.html) with HTTP proxy settings (while every other non-proxied GCP call is blocked) could lead to this issue.

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by arteam**
Currently, if the user set proxy setting for the GCS repository, they are only used for interactions with the GCS API, but not the other network interactions that the the module performs (refreshing OAuth tokens, fetching metadata) outside of the SDK.

Fix that by making sure that all network interactions go via proxy if it's configured.

Fixes #91952

### CODE REVIEW

**Comment by arteam** [comment]
@elasticmachine update branch

File: docs/changelog/92192.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+pr: 92192
+summary: Support for GCS proxies everywhere in the GCS API
```
Comment (pxsalehi) [inline docs/changelog/92192.yaml:2]: nit: reading this, I find GCS proxy might be a bit misleading. Maybe just support proxies?

File: modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClientSettingsTests.java [hunk 2]
```diff
@@ -139,14 +139,72 @@ public void testLoadsProxySettings() throws Exception {
...
+                clientName,
+                Base64.getEncoder().encodeToString(KeyPairGenerator.getInstance("RSA").generateKeyPair().getPrivate().getEncoded()),
+                clientName,
+                clientName,
+                URI.create("http://oauth2.googleapis.com/oauth2/token")
+            ).getBytes(StandardCharsets.UTF_8)
         );
+        var settings = Settings.builder().setSecureSettings(secureSettings).build();
+        var proxyServerSocket = new MockServerSocket(0);
+        var proxyServerThread = new Thread(() -> {
```
Comment (pxsalehi) [inline modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClientSettingsTests.java:173]: We're bringing this proxy up and run a thread to process the request, in two places now. Would it be useful to DRY it and make a nice test utility out of this?

File: modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClientSettingsTests.java [hunk 3]
```diff
@@ -139,14 +139,72 @@ public void testLoadsProxySettings() throws Exception {
...
+                            "access_token": "proxy_access_token",
+                            "token_type": "bearer",
+                            "expires_in": 3600
+                        }
+                        """;
+                    socket.getOutputStream().write(formatted("""
+                        HTTP/1.1 200 OK\r
+                        Content-Length: %s\r
+                        \r
+                        %s\r
```
Comment (pxsalehi) [inline modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClientSettingsTests.java:191]: Why do we need the `\r`s here? Test seems to pass also w/o them.

**Review by pxsalehi — APPROVED** [review]
Thanks Artem for fixing this. Nice that you have the tests that actually start a proxy! I just left a few optional suggestions and comments. LGTM!

**Review by DaveCTurner — COMMENTED** [review]
Yes I think this is worth backporting to 8.6 - it's an annoying bug for proxy users, and the production code change doesn't look too controversial.

The testing is nice, but I would like to have more of an end-to-end test too - i.e. one which sets up the repo and then does some blob operations on it, to validate that a repo with a proxy config really _only_ talks via the proxy. Would that be possible?

File: modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClientSettingsTests.java [hunk 4]
```diff
@@ -139,14 +139,72 @@ public void testLoadsProxySettings() throws Exception {
...
+        });
+        proxyServerThread.start();
+        var proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(InetAddress.getLoopbackAddress(), proxyServerSocket.getLocalPort()));
+
+        try {
+            ServiceAccountCredentials credentials = loadCredential(settings, clientName, proxy);
+            assertNotNull(credentials);
+            assertEquals("proxy_access_token", SocketAccess.doPrivilegedIOException(credentials::refreshAccessToken).getTokenValue());
+        } finally {
+            proxyServerThread.interrupt();
```
[the remark this answers is not in the document]
Comment (arteam) [reply modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClientSettingsTests.java]: Actually, closing the server socket triggers `SocketException` in the loop, and allows the thread to exit. So, `join` isn't needed here.

**Comment by arteam** [comment]
> The testing is nice, but I would like to have more of an end-to-end test too - i.e. one which sets up the repo and then does some blob operations on it, to validate that a repo with a proxy config really only talks via the proxy. Would that be possible?

I've added an integration test `GcsProxyIntegrationTests` that routes all requests via a web proxy server that sets a `X-Via` header, so we can verify that all requests come via the proxy.

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Review by pxsalehi — APPROVED** [review]

**Comment by DaveCTurner** [comment]
@pxsalehi are you ok with backporting this all the way to 8.6.2? I haven't looked at it in detail but this seems like quite a big change to make to a released branch. Maybe 8.7.0 is enough?

**Comment by pxsalehi** [comment]
Also technically we're past feature freeze for 8.7.0. I think. Should this go to 8.7.1?

**Comment by DaveCTurner** [comment]
It doesn't really matter, the labels are adjusted on release. I expect this'll land in 8.7.0 tho.

**Comment by arteam** [comment]
@elasticmachine update branch

**Comment by arteam** [comment]
@elasticmachine update branch

**Pull request merged**